### ISSUE TITLE
Naming conventions for rules

### ISSUE THREAD

**Issue by annethyme** [act-rules/act-rules.github.io#250]
Skimming over our list of published rules on https://auto-wcag.github.io/auto-wcag/pages/rules.html, there are big differences in the naming. Singular or plural, with or without dashes, wordy or to the bone, with or without verbs. 
Now that we have tried different things out for a while, I suggest we discuss and:
- decide on a best practice
- write down the best practice
- align the existing rules, PRs and issues to the agreement we make

**Comment by WilcoFiers** [comment]
Some thoughts:

- Front load rule names so that appropriate element names / roles go first: "button has accessible name" instead of "accessible name for button"
- Use singular tense: "button has accessible name" instead of "buttons have an accessible name"
- Describe what passes, not what fails: "button has accessible name" instead of "button without accessible name"
- ...?

**Comment by WilcoFiers** [comment]
Moving this off the September agenda. We need to get open PRs resolved. We'll pick it up in October.

**Comment by annethyme** [comment]
How about these?

- "button has accessible name" or "button has *an* accessible name"?
- Capitalization? "*b*utton has accessible name" or "*B*utton has accessible name"
- "iFrame has name" or "iframe has name" or "Iframe has name"

**Comment by WilcoFiers** [comment]
| Old name                                   | New name      |
|--------------------------------------------|---------------|
| Old rule name                              | New rule name |
| ARIA Attribute Is Valid                    | ARIA attributes are valid |
| ARIA State Or Property Allowed             | ARIA attributes are allowed for role |
| ARIA State Or Property Has Valid Value     | ARIA attribute values are valid |
| Audio-Only As A Media Alternative For Text | audio elements are a media alternative |
| Autocomplete Valid                         | Autocomplete attributes are valid |
| Buttons Have An Accessible Name            | buttons have an accessible name |
| Form Field Has Accessible Name             | form fields have an accessible name |
| Form Field Label Is Descriptive            | form field labels are descriptive |
| HTML Page Has A Title                      | html page has a title |
| HTML Has Lang Attribute                    | html element has a lang attribute |
| HTML Lang And Xml:lang Match               | html element has matching lang and xml:lang attribute |
| Heading Is Descriptive                     | headings are descriptive |
| Id Attribute Is Unique                     | id attributes are unique |
| Image Has Accessible Name                  | images have an accessible name |
| Links Have An Accessible Name              | links have an accessible name |
| Meta-Refresh No Delay                      | meta element does not have a delayed redirect |
| No Keyboard Trap                           | keyboard focus is not trapped |
| No Keyboard Trap Non-Standard Navigation   | keyboard traps have non-standard escape |
| No Keyboard Trap Standard Navigation       | keyboard traps have standard escape |
| Role Attribute Has Valid Value             | role attribute values are valid |
| Role Has Required States And Properties    | ARIA roles have required ARIA attributes |
| Valid Body Lang Attribute                  | lang attribute values of content are valid |
| Validity Of HTML Lang Attribute            | html element lang attribute value is valid |
| Video Element Description Track            | video elements have description track |
| Video-Only As A Media Alternative For Text | video elements without audio are a media alternative |
| Vídeo Has Audio Alternative                | video elements have audio alternative |
| Aria-Hidden With Focusable Content         | ARIA hidden content is not focusable |
| Attributes Are Not Duplicated              | attributes are not duplicated |
| Audio Described Video Element              | video elements have audio description |
| Audio Elements Captions                    | audio elements have captions |
| Iframe Has An Accessible Name              | iframes have an accessible name |
| Label And Name From Content Mismatch       | name from content included in accessible name |
| Video Element Transcript                   | video elements have transcript |
| Video Has Captions                         | video elements have captions |

**Comment by EmmaJP** [comment]
I understand why ARIA is capitalised, but is Autocomplete a proper noun?

**Comment by WilcoFiers** [comment]
@annethyme @EmmaJP The reason I went with plural is because rules are written in plural. They say things like "all element that.. " and "each target node". As for capitalisation, I'll fix that.

**Comment by jkodu** [comment]
@WilcoFiers is there still work to be done on these?

**Comment by annethyme** [comment]
On plural vs. singular: 

I find it weird that is says "buttons have (plural) an accessible name (singular)".
If we want to go plural, it seems more logical to me to say "buttons have (plural) accessible names (plural)". 
Because we don't want all the buttons to share the same accessible name, which is not really clear from "buttons have an accessible name", if that is what we mean. And there could actually be future rules where we would want to check for a many-to-one relationship, which wouldn't be clearly distinguishable if we by default has a many-to-one wording for all rules.

And while the rule applies to e.g. "any button in the test target", the Expectation is usually something like "Each button has an accessible name". So I think using singular for describing what condition that we are checking for makes more sense.
Basically we are running the "button has an accessible name" condition on all the test targets of the page.

I definitely vote for singular: "Button has an accessible name".

**Comment by annethyme** [comment]
Another concern of mine related to naming of rules:

These two rules are related, but that is not at all clear from the terms used in their naming:
- HTML page has a title
- Document title is descriptive 

I think they should be using the same wording, either "HTML page" or "document", so that it is clear that they are checking different things around the same element. 

Which one should it be here?

It doesn't seem that either of the terms are (yet) used in other rule names.

**Comment by EmmaJP** [comment]
This issue has been going for a while, so thought I'd summarise where I think it's got to.

The intention / need is to decide and document a best practice for rule names / titles. And then to make the necessary updates.

From the conversation above, the following points have come out:

- singular or plural: the preference is leaning to singular tense so that necessary plurals stand out, for example 'video has captions'
- with or without dashes: without dashes, as no longer affecting URL, unless correct for code, eg. my-input custom element
- wordy or succinct: direct, declarative, and _possibly_ without unnecessary words like 'an'
- with or without verbs: the preference is with verbs, declarative (eg has)
- capitalization: sentence case, eg. Button has accessible name, and initials caps, eg. SVG
- camelcase: only where correct for code, eg. iframe not iFrame, feImage in SVG
- front load applicable element: eg. Button has accessible name
- consistent: refer to applicable element in a consistent way across rules, especially when related or checking the same applicable element
- describe passing condition: eg. id value is unique, rather than page has no duplicate ids
- refer to WCAG language used
- ideally unique: no 2 rules should have same title, this should happen anyway if title includes the applicable element and passing condition

**Comment by WilcoFiers** [comment]
| Old name | New name |
|----------|----------|
| ARIA attribute is valid | aria-* attribute is defined in WAI-ARIA |
| ARIA state or property allowed | ARIA state or property permitted |
| ARIA state or property has valid value | ARIA state or property has valid value |
| Attributes are not duplicated | Attribute is not duplicated |
| Id attribute is unique | id attribute value is unique |
| Audio only has a text alternative | audio element content has text alternative |
| Audio elements have a transcript | audio element has transcript |
| Audio-only as a media alternative for text | audio element is media alternative for text |
| Video as a media alternative for text | video element is media alternative for text  |
| Video element audio described | video element has audio description |
| Video element description track | video element has description track |
| Video element transcript | video element has transcript |
| Video has audio alternative | video element has alternative for audio |
| Video has captions | video element has captions |
| Video only element has description track | video element without audio has description track |
| Video only element has transcript | video element without audio has transcript |
| Video only has an accessible alternative | video element without audio has an accessible alternative |
| Video with audio has audio description | video element has audio description |
| Video with audio has audio descriptions or transcript | video element has audio description or transcript |
| Video-only as a media alternative for text | video element without audio is media alternative for text |

**todo**:

- Autocomplete valid
- Buttons have an accessible name
- Document title is descriptive
- Filename is valid accessible name
- Form field has accessible name
- Form field label is descriptive
- HTML Page has a title
- HTML has lang attribute
- HTML lang and xml:lang match
- Heading is descriptive
- Iframe has an accessible name
- Iframes with identical accessible names serve equivalent purpose
- Image button has accessible name
- Image has accessible name
- Label and name from content mismatch
- Links have an accessible name
- Links with identical accessible names serve equivalent purpose
- Meta-refresh no delay
- No keyboard trap
- No keyboard trap non-standard navigation
- No keyboard trap standard navigation
- Role attribute has valid value
- Role has required states and properties
- Valid body lang attribute
- Validity of HTML Lang attribute
- aria-hidden with focusable content
- video only has audio alternative

**Comment by EmmaJP** [comment]
| Old name | New name |
|----------|----------|
| Autocomplete valid | autocomplete attribute has valid value |
| Buttons have an accessible name | Button has accessible name |
| Document title is descriptive | HTML page title is descriptive |
| Filename is valid accessible name | Image filename is valid accessible name |
| Form field has accessible name | - |
| Form field label is descriptive | - |
| HTML Page has a title | HTML page has valid title |
| HTML has lang attribute | html element has valid lang attribute |
| HTML lang and xml:lang match | lang and xml:lang attributes on element have matching values |
| Heading is descriptive | - |
| Iframe has an accessible name | iframe element has accessible name |
| Iframes with identical accessible names serve equivalent purpose | iframe elements with identical accessible names serve equivalent purpose |
| Image button has accessible name | - |
| Image has accessible name | - |
| Label and name from content mismatch |  Visible label is part of accessible name |
| Links have an accessible name | Link has accessible name |
| Links with identical accessible names serve equivalent purpose | - |
| Meta-refresh no delay | meta element has no refresh delay |
| No keyboard trap | HTML or SVG element has no keyboard trap |
| No keyboard trap non-standard navigation | ? |
| No keyboard trap standard navigation | ? |
| Role attribute has valid value | role attribute has valid value |
| Role has required states and properties | Element with role has required states and properties |
| Valid body lang attribute | Element within body has valid lang attribute |
| Validity of HTML Lang attribute | lang attribute value is permitted |
| aria-hidden with focusable content | Element with aria-hidden has no focusable content |

**Comment by EmmaJP** [comment]
Feedback on above suggestions from myself and Wilco much appreciated.

**Comment by jkodu** [comment]
@EmmaJP ones that have been worked upon by yourself and @WilcoFiers look good. Nitpick, seems like everything is kept to lower casing in this [comment](https://github.com/act-rules/act-rules.github.io/issues/250#issuecomment-521980486). Something to update on your listing above.

Also are you looking for opinion on the dashes and question mark ones?

**Comment by EmmaJP** [comment]
@jkodu refer you to stated preference...

capitalization: sentence case unless element code, eg. Heading is descriptive, eg. aria-* attribute is defined in specification, and initials caps, eg. SVG

We've discussed this, and doing something more to indicate code elements and attributes, such as back ticks.

**Comment by EmmaJP** [comment]
Suggested wording for rule-design page:

**Rule name and identifier**

The rule should have a unique name, and will have a unique identifier. The 6-digit identifier is created using a random hex generator, such as https://www.browserling.com/tools/random-hex. The conventions for naming of rules are as follows:

- Be succinct, direct and declarative, and avoid unnecessary words, eg. an, the.
- Use declarative verbs, eg. has, have, is.
- Use back ticks to indicate words from code, eg. `label` v. Label.
- Identify code as an element or attribute, eg. `label` element.
- Use sentence case, eg. Heading is descriptive, unless using code, eg. `id` attribute value is unique, or initials, eg. HTML.
- Use singular tense so that necessary plurals stand out, eg. `video` element has captions.
- Front load the applicable thing, eg. Button has accessible name.
- Refer to each applicable thing in a consistent way across rules, especially when related or checking the same applicable thing.
- Describe the passing condition, eg. `id` value is unique, rather than page has no duplicate ids.
- Do not use hyphens or dashes unless correct for code, eg. `aria-*`.
- Do not use camelCase unless correct for code, eg. `iframe` not iFrame, `feImage` in SVG.
- If unsure, refer to WCAG language used.
- Aim to be unique, which should happen if title includes the applicable thing and passing condition.

**Comment by EmmaJP** [comment]
Yikes - sorting out the video stuff was scary!

**Comment by jkodu** [comment]
Draft PR #946 in place to update rule design.
